lock-api/src/main/java/com/palantir/lock/watch/LockWatchStateUpdater.java

```diff
@@ -0,0 +1,77 @@
+/*
+ * (c) Copyright 2019 Palantir Technologies Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.lock.watch;
+
+import java.util.Map;
+import java.util.OptionalLong;
+import java.util.Set;
+import java.util.UUID;
+import java.util.stream.Collectors;
+
+import com.google.common.collect.RangeSet;
+import com.palantir.lock.LockDescriptor;
+
+public class LockWatchStateUpdater implements LockWatchEvent.Visitor<Void> {
+    private final RangeSet<LockDescriptor> watches;
+    private final Map<LockDescriptor, LockWatchInfo> lockWatchState;
+    private final Set<UUID> unmatchedOpenLocksEvents;
+
+    LockWatchStateUpdater(
+            RangeSet<LockDescriptor> watches,
+            Map<LockDescriptor, LockWatchInfo> lockWatchState,
+            Set<UUID> unmatchedOpenLocksEvents) {
+        this.watches = watches;
+        this.lockWatchState = lockWatchState;
+        this.unmatchedOpenLocksEvents = unmatchedOpenLocksEvents;
+    }
+
+    @Override
+    public Void visit(LockEvent lockEvent) {
+        for (LockDescriptor descriptor : lockEvent.lockDescriptors()) {
+            lockWatchState.put(descriptor, LockWatchInfo.of(LockWatchInfo.State.LOCKED, lockEvent.sequence()));
+        }
+        return null;
+    }
+
+    @Override
+    public Void visit(UnlockEvent unlockEvent) {
+        for (LockDescriptor descriptor : unlockEvent.lockDescriptors()) {
+            OptionalLong lastLocked = lockWatchState.getOrDefault(descriptor, LockWatchInfo.NOT_WATCHED).lastLocked();
+            lockWatchState.put(descriptor, LockWatchInfo.of(LockWatchInfo.State.UNLOCKED, lastLocked));
+        }
+        return null;
+    }
+
+    @Override
+    public Void visit(LockWatchOpenLocksEvent openLocksEvent) {
+        for (LockDescriptor descriptor : openLocksEvent.lockDescriptors()) {
+            lockWatchState.put(descriptor, LockWatchInfo.of(LockWatchInfo.State.LOCKED, openLocksEvent.sequence()));
+        }
+        unmatchedOpenLocksEvents.add(openLocksEvent.lockWatchId());
+        return null;
+    }
+
+    @Override
+    public Void visit(LockWatchCreatedEvent lockWatchCreatedEvent) {
```

Comment:
Need to check that the obvious race condition works:
- there is an open-locks event, associated with watch ID `Q`
- then there is a lock locked that is related to `Q`
- then we get the created event

I guess we still have it in the Info, even if we aren't sure how it ended up there at that point, so that looks fine. Would like to sanity check this thing next time we sync up though.

Reply:
Yes, that's the intended behaviour -- having open locks means we didn't miss any currently held locks, then we register the lock watch created event, but maybe some of the locks have been taken in the meantime -- they will, however, be logged, and we will correctly identify that they are in fact locked once the lock watch created event is processed.

The purpose of the LWCE is to let us know that everything in the lock watch request that is not explicitly locked is in fact not locked. Added a test for the case you mention.